service/history/archivalQueueTaskExecutor.go

```diff
@@ -0,0 +1,208 @@
+	isActive bool, err error) {
+	task := executable.GetTask()
+	taskType := queues.GetArchivalTaskTypeTagValue(task)
+	tags = []metrics.Tag{
+		getNamespaceTagByID(e.shardContext.GetNamespaceRegistry(), task.GetNamespaceID()),
+		metrics.TaskTypeTag(taskType),
+	}
+	switch task := task.(type) {
+	case *tasks.ArchiveExecutionTask:
+		err = e.processArchiveExecutionTask(ctx, task)
+	default:
+		err = fmt.Errorf("task with invalid type sent to archival queue: %+v", task)
+	}
+	return tags, true, err
+}
+
+func (e *archivalQueueTaskExecutor) processArchiveExecutionTask(ctx context.Context, task *tasks.ArchiveExecutionTask) (err error) {
+	weContext, release, err := getWorkflowExecutionContextForTask(ctx, e.workflowCache, task)
+	if err != nil {
+		return err
+	}
+	defer func() { release(err) }()
+
+	mutableState, err := weContext.LoadMutableState(ctx)
+	if err != nil {
+		return err
+	}
+	if mutableState == nil || mutableState.IsWorkflowExecutionRunning() {
+		return nil
+	}
+	lastWriteVersion, err := mutableState.GetLastWriteVersion()
+	if err != nil {
+		return err
+	}
+	namespaceEntry := mutableState.GetNamespaceEntry()
+	err = CheckTaskVersion(
+		e.shardContext,
+		e.logger,
+		namespaceEntry,
+		lastWriteVersion,
+		task.Version,
+		task,
+	)
+	if err != nil {
+		return err
+	}
+
+	namespaceName := namespaceEntry.Name()
+	nextEventID := mutableState.GetNextEventID()
+	executionInfo := mutableState.GetExecutionInfo()
+	workflowTypeName := executionInfo.GetWorkflowTypeName()
+
+	executionState := mutableState.GetExecutionState()
+	memo := getWorkflowMemo(copyMemo(executionInfo.Memo))
+	closeTime, err := mutableState.GetWorkflowCloseTime(ctx)
+	if err != nil {
+		return err
+	}
+	if closeTime == nil {
+		return errors.New("can't archive workflow with nil close time")
+	}
+	branchToken, err := mutableState.GetCurrentBranchToken()
+	if err != nil {
+		return err
+	}
+
+	var targets []archival.Target
+	if e.shardContext.GetArchivalMetadata().GetVisibilityConfig().ClusterConfiguredForArchival() &&
+		namespaceEntry.VisibilityArchivalState().State == enumspb.ARCHIVAL_STATE_ENABLED {
+		targets = append(targets, archival.TargetVisibility)
+	}
+	if e.shardContext.GetArchivalMetadata().GetHistoryConfig().ClusterConfiguredForArchival() &&
+		namespaceEntry.HistoryArchivalState().State == enumspb.ARCHIVAL_STATE_ENABLED {
+		targets = append(targets, archival.TargetHistory)
+	}
+
+	historyURI, err := carchiver.NewURI(namespaceEntry.HistoryArchivalState().URI)
+	if err != nil {
+		e.logger.Error(
+			"Failed to parse history URI.",
+			tag.ArchivalURI(namespaceEntry.HistoryArchivalState().URI),
+			tag.Error(err),
+		)
+		return nil
+	}
+	visibilityURI, err := carchiver.NewURI(namespaceEntry.VisibilityArchivalState().URI)
+	if err != nil {
+		e.logger.Error(
+			"Failed to parse visibility URI.",
+			tag.ArchivalURI(namespaceEntry.VisibilityArchivalState().URI),
+			tag.Error(err),
+		)
+		return nil
+	}
+	_, err = e.archiver.Archive(ctx, &archival.Request{
```

Comment:
I spoke with @yux0 about this, and we shared the concern. However, YX said it would be OK since we're guaranteed the workflow is closed, so nothing else should access this workflow unless we revive it. Can we acquire the lock to get the attributes we need, release it when archiving, and then reacquire it when we add the follow-up tasks?

Reply:
This is now fixed